lib/stages/oneDiskVDIFWrite.cpp

```diff
@@ -0,0 +1,338 @@
+    auto& tel = Telescope::instance();
+    time_res_nsec = tel.seq_length_nsec();
+    // Set up buffers
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+}
+
+oneDiskVDIFWrite::~oneDiskVDIFWrite(){}
+
+
+void oneDiskVDIFWrite::main_thread(){
+    int fd;
+    int payload_ctr;
+    uint8_t* input = nullptr;
+    frameID in_frame_id(in_buf);
+    uint8_t* vdif_frame;
+    VDIFHeader* vdif_header; // VDIF header
+    uint8_t* payload;  // VDIF payload
+    bool isFileOpen = false;
+    char data_time[64];
+    char data_set_c[150];
+    char data_dir_name[256];
+    const int full_path_len = 512;
+    char full_path[full_path_len];    
+    time_t rawtime;
+    struct tm* timeinfo;
+    size_t file_num;
+    uint32_t time_offset = 0;
+    uint32_t freq_offset = 0;
+    uint32_t vdif_frame_data_size;
+    uint32_t vdif_frame_size;
+    uint32_t nvdif_payload_per_file;
+    uint32_t in_buf_frame_size;
+    uint32_t src_start;
+    uint32_t target_start;
+    uint32_t inframe_ctr = 0;
+    long intsec_offset;
+    char setting_log_file[full_path_len];
+    char data_log_file[full_path_len];
+    FILE* setting_file;
+    FILE* j_file;
+
+    MergedBeamMetadata* in_metadata = nullptr;
+    timespec out_frame_unix_time;
+    timespec vdif_file_start_time;
+
+
+    time(&rawtime);
+
+    timeinfo = gmtime(&rawtime);
+    // The time prescision in the file should never be more than a second. 
+    // Since creating the directory is independent for each disk now.
+    // To make sure the directory name is the same, the precision of time stamp
+    // in the directory name should be limited to seconds.
+    strftime(data_time, sizeof(data_time), "%Y%m%dT%H%M%SZ", timeinfo);
+    snprintf(data_set_c, sizeof(data_set_c), "%s_%s_%s", data_time, instrument_name.c_str(),
+             file_ext.c_str());
+    
+    // Make output directory
+    if (write_to_disk) {
+        if (use_abs_path){
+            snprintf(data_dir_name, sizeof(data_dir_name), "%s", abs_path.c_str());
+	}
+	else{
+	    snprintf(data_dir_name, sizeof(data_dir_name), "%s/%s/%d/%s", disk_base.c_str(), disk_set.c_str(), disk_id, data_set_c);
+	}
+	make_dir(data_dir_name);
+	// write setting log
+	snprintf(setting_log_file, full_path_len, "%s/%s.txt", data_dir_name, "setting_log");
+        setting_file = fopen(setting_log_file, "w");
+	snprintf(data_log_file, full_path_len, "%s/%s.txt", data_dir_name, "data_log");
+	j_file = fopen(data_log_file, "w");
+    }
+    
+        
+    // allocate the vdif buffer 
+    vdif_frame_data_size = vdif_samples_per_frame * vdif_freq_per_frame * num_pol;
+    vdif_frame_size = vdif_frame_header_size + vdif_frame_data_size;
+    INFO("VDIF size {:d} {:d}", vdif_frame_data_size, vdif_frame_size);
+    vdif_frame = (uint8_t*) malloc(vdif_frame_size);
+    // Cast vdif header
+    vdif_header = (VDIFHeader*) &vdif_frame[0];
+    // cast vdif payload
+    payload = &vdif_frame[vdif_frame_header_size];
```

Comment:
You could also use pointer math.  Although one could argue it's clearer to read with an array access.
```suggestion
    payload = vdif_frame + vdif_frame_header_size;
```